### PR TITLE
Support for sharing snapshots

### PR DESCRIPTION
**Pull request by ambarve**
Contents in different namespaces can be shared by setting the content sharing policy. This
change adds the support for doing the same for snapshots. If same image is pulled and
extracted in different namespaces the same image layers will be extracted in multiple
namespaces. The snapshots created for these extracted layers contain the exact same
content. To avoid this duplication of work in improve the performance of image pull (if
the same image is already pulled in another namespace) snapshots of existing layers can be
shared across namespaces. Setting the newly added snapshot sharing policy will enable this.

Adding following to containerd.toml should enable this.

```
  [plugins.bolt]
        snapshot_sharing_policy = "shared"
```

Signed-off-by: Amit Barve <ambarve@microsoft.com>

### CODE REVIEW

**Comment by ambarve** [comment]
This PR also cherry-picks the commits for improving parallel image pull performance & sharing content between namespaces as they are required to build this branch.

File: image.go [hunk 1]
```diff
@@ -355,7 +355,7 @@ func (i *image) Unpack(ctx context.Context, snapshotterName string, opts ...Unpa
 
 	for _, layer := range layers {
 		unpacked, err = rootfs.ApplyLayerWithOpts(ctx, layer, chain, sn, a, config.SnapshotOpts, config.ApplyOpts)
-		if err != nil {
+		if err != nil && !errdefs.IsAlreadyExists(err) {
```
Comment (helsaawy) [inline image.go:371]: Is there a short circuit way to check if the layer already exists besides trying to apply it?
Or is this the easiest/most straight forward way?

File: snapshots/snapshotter.go [hunk 2]
```diff
@@ -35,6 +38,19 @@ const (
...
+// snapshot sharing is enabled and if the target snapshot (i.e the final digest when the snapshot is
+// being committed) already exists, the Prepare call returns this error to let the client know that
+// the target snapshots exists. This can only happen during image extraction scenarios where you
+// already know the final digest after the layer contents are written into the snapshot.  Since this
+// error wraps the ErrAlreadyExists, when checking for errors first `IsTargetSnapshotExists` should
+// be checked.
+var ErrorTargetSnapshotAlreadyExists = fmt.Errorf("target snapshot already exists: %w", errdefs.ErrAlreadyExists)
+
+func IsTargetSnapshotExists(e error) bool {
+	return errors.Is(e, errdefs.ErrAlreadyExists) && strings.HasPrefix(e.Error(), "target snapshot already exists")
```
Comment (helsaawy) [inline snapshots/snapshotter.go:51]: `errors.Is(e, errdefs.ErrAlreadyExists)` is equivalent to to `errdefs.IsAlreadyExists(e)`, right?
can we use the latter here to keep it consistent?

File: unpacker.go [hunk 3]
```diff
@@ -158,7 +158,7 @@ EachLayer:
 			key = fmt.Sprintf(snapshots.UnpackKeyFormat, uniquePart(), chainID)
 			mounts, err = sn.Prepare(ctx, key, parent.String(), opts...)
 			if err != nil {
-				if errdefs.IsAlreadyExists(err) {
+				if errdefs.IsAlreadyExists(err) || snapshots.IsTargetSnapshotExists(err) {
```
Comment (helsaawy) [inline unpacker.go:169]: The `errdefs.IsAlreadyExists(err)` bit should be redundant with the check `IsTargetSnapshotExists`, right?

File: metadata/snapshot.go [hunk 4]
```diff
@@ -307,6 +345,8 @@ func (s *snapshotter) createSnapshot(ctx context.Context, key, parent string, re
 		bopts   = []snapshots.Opt{
 			snapshots.WithLabels(snapshots.FilterInheritedLabels(base.Labels)),
 		}
+		targetPresentInNamespace string
```
Comment (helsaawy) [inline metadata/snapshot.go:348]: `targetPresentInNamespace` reads like a `bool` value, can this be renames to `targetNamespace` or something similar?

File: metadata/snapshot.go [hunk 5]
```diff
@@ -280,6 +282,42 @@ func (s *snapshotter) View(ctx context.Context, key, parent string, opts ...snap
...
+
+	if checkAllNamespaces {
+		vbkt := tx.Bucket(bucketKeyVersion)
+		if vbkt == nil {
+			return false, "", fmt.Errorf("can not find version bucket")
+		}
+
+		// iterate through each namespace
+		vc := vbkt.Cursor()
+		for nk, _ := vc.First(); nk != nil; nk, _ = vc.Next() {
```
Comment (helsaawy) [inline metadata/snapshot.go:307]: It feels redundant to search the same namespace here as we did on 295-297 when we iterate through all of them

File: snapshots/snapshotter.go [hunk 6]
```diff
@@ -35,6 +38,19 @@ const (
...
 )
 
+// ErrTargetSnapshotAlreadyExists is different than ErrAlreadyExists when it comes to snapshots.  If
+// snapshot sharing is enabled and if the target snapshot (i.e the final digest when the snapshot is
+// being committed) already exists, the Prepare call returns this error to let the client know that
+// the target snapshots exists. This can only happen during image extraction scenarios where you
+// already know the final digest after the layer contents are written into the snapshot.  Since this
+// error wraps the ErrAlreadyExists, when checking for errors first `IsTargetSnapshotExists` should
+// be checked.
+var ErrorTargetSnapshotAlreadyExists = fmt.Errorf("target snapshot already exists: %w", errdefs.ErrAlreadyExists)
```
Comment (helsaawy) [inline snapshots/snapshotter.go:48]: `ErrTargetSnapshotAlreadyExists`, or, better yet, `ErrSnapshotAlreadyExists`

File: services/server/server.go [hunk 7]
```diff
@@ -432,19 +433,35 @@ func LoadPlugins(ctx context.Context, config *srvconfig.Config) ([]*plugin.Regis
 				snapshotters[name] = sn.(snapshots.Snapshotter)
 			}
 
-			shared := true
-			ic.Meta.Exports["policy"] = srvconfig.SharingPolicyShared
+			contentShared := true
+			ic.Meta.Exports["content_policy"] = srvconfig.SharingPolicyShared
```
Comment (helsaawy) [inline services/server/server.go:437]: are the `Exports` anywhere, or just when running `ctr plugins ls -d`?
also, minor nit, should it be `contentPolicy`, to be more go-like?

**Review by helsaawy — COMMENTED** [review]
Overall lgtm, just minor comments here and there

**Comment by ambarve** [comment]
Closing this as this is a complicated change that requires further discussion.

**Pull request closed without merging**